mpc/garble/mpc-garble-core/src/generator.rs

```diff
@@ -37,73 +55,182 @@ pub(crate) fn and_gate<C: BlockCipher<BlockSize = U16> + BlockEncrypt>(
+                circ.inputs().len(),
+                inputs.len(),
+            ))?;
+        }
+
+        let mut low_labels: Vec<Option<Label>> = vec![None; circ.feed_count()];
+        for (encoded, input) in inputs.iter().zip(circ.inputs()) {
+            if encoded.value_type() != input.value_type() {
+                return Err(TypeError::UnexpectedType {
+                    expected: input.value_type(),
+                    actual: encoded.value_type(),
+                })?;
             }
-            Gate::Xor {
-                xref, yref, zref, ..
-            } => {
-                let x = labels[xref].ok_or(EncodingError::UninitializedLabel(xref))?;
-                let y = labels[yref].ok_or(EncodingError::UninitializedLabel(yref))?;
-                let z = xor_gate(&x, &y, delta);
-                labels[zref] = Some(z);
+
+            for (label, node) in encoded.iter().zip(input.iter()) {
+                low_labels[node.id()] = Some(*label);
             }
-            Gate::And {
-                xref, yref, zref, ..
-            } => {
-                let x = labels[xref].ok_or(EncodingError::UninitializedLabel(xref))?;
-                let y = labels[yref].ok_or(EncodingError::UninitializedLabel(yref))?;
-                let (z, t) = and_gate(cipher, &x, &y, delta, gid);
-                encrypted_gates.push(t);
-                labels[zref] = Some(z);
-                gid += 2;
+        }
+
+        Ok(Self {
+            cipher: Aes128::new_from_slice(&CIPHER_FIXED_KEY).expect("cipher should initialize"),
+            circ,
+            delta,
+            low_labels,
+            pos: 0,
+            gid: 1,
+            hasher: if hash { Some(Hasher::new()) } else { None },
+        })
+    }
+
+    /// Returns whether the generator has finished generating the circuit.
+    pub fn is_complete(&self) -> bool {
+        self.pos >= self.circ.gates().len()
+    }
+
+    /// Returns the encoded outputs of the circuit.
+    pub fn outputs(&self) -> Result<Vec<EncodedValue<state::Full>>, GeneratorError> {
+        if !self.is_complete() {
+            return Err(GeneratorError::NotFinished);
+        }
+
+        Ok(self
+            .circ
+            .outputs()
+            .iter()
+            .map(|output| {
+                let labels: Vec<Label> = output
+                    .iter()
+                    .map(|node| self.low_labels[node.id()].expect("feed should be initialized"))
+                    .collect();
+
+                EncodedValue::<state::Full>::from_labels(output.value_type(), self.delta, &labels)
+                    .expect("encoding should be correct")
+            })
+            .collect())
+    }
+
+    /// Returns the hash of the encrypted gates.
+    pub fn hash(&self) -> Option<Hash> {
+        self.hasher.as_ref().map(|hasher| {
+            let hash: [u8; 32] = hasher.finalize().into();
+            Hash::from(hash)
+        })
+    }
+}
+
+impl Iterator for Generator {
```

Comment:
I am not sure if it is a good idea to have state-modifying behavior inside an iterator implementation, which has nothing to do with iteration. You feed things into the hasher and modify `gid` while iterating.

In contrast, the evaluator does these things in `evaluate`. Maybe it is better to have a special method which does these things for the generator as well.

Reply:
I think my thought was, that if the `Generator` is used without iteration somehow, then the hash is wrong.

Reply:
Hmm I see. Ok then let's keep it and maybe add a note?